Add support for FASP data sharing

oneiros opened the pull request:
This implements https://github.com/mastodon/fediverse_auxiliary_service_provider_specifications/blob/main/discovery/data_sharing/v0.1/data_sharing.md

Discovery providers can subscribe to "lifecycle events" of both accounts and statuses. They then get notified whenever a (discoverable) account or an (indexable and public) status is created, updated or deleted.

Discovery providers can also subscribe to be informed of events that could signal a trend (e.g. more than 5 likes within 10 minutes).

I implemented all of this using `after_commit` callbacks on the respective models. This might not be the best way, but it seemed to be the easiest.

It is worth noting, that this operates on individual accounts and statuses, while the API would allow to batch them. But I am not sure if the added complexity of buffering data before sending is worth it.

Lastly, there is also "backfill requests" where the provider asks to learn of existing accounts and/or statuses.

oneiros committed:
Humble beginnings of fasp data sharing.

Subscriptions can be made and new statuses will be announced to
subscribed fasp.

oneiros committed:
Handle full content lifecycle for subscribed fasps

oneiros committed:
Add account and trend data sharing

oneiros committed:
Select what to share with fasp

Only share statuses where the account has `#indexable` set
to `true`.

Only share accounts where `#discoverable` is set to `true`, with
one exception: If `#discoverable` has just been set to `false`
this is an important information for the fasp.

oneiros committed:
Fix typo

oneiros committed:
Only share public statuses.

Only actually public, i.e. not "unlisted". One exception: Share
an update if the update changed the visibility from "public" to
something else. The fasp can act on this information then.

oneiros committed:
Implement backfill requests

oneiros committed:
Fix autogenerated fabricators

oneiros committed:
Put FASP support behind a feature flag

oneiros committed:
Add model specs

oneiros committed:
Add request specs

oneiros committed:
Add specs for model hooks

oneiros committed:
Fix broken specs

Rubocop fooled me into breaking this.

oneiros committed:
Add worker specs

oneiros committed:
Improve comment to explain redundant call

ClearlyClaire commented on app/models/fasp/subscription.rb:
```diff
@@ -0,0 +1,43 @@
+  validates :subscription_type, presence: true,
+                                inclusion: TYPES
+
+  scope :content, -> { where(category: 'content') }
+  scope :account, -> { where(category: 'account') }
+  scope :lifecycle, -> { where(subscription_type: 'lifecycle') }
+  scope :trends, -> { where(subscription_type: 'trends') }
```
Maybe those scopes could use some renaming. `Fasp::Subscription.account` was a bit confusing, as on first read I did not think this was a scope, but an Account attribute.

ClearlyClaire replied:
I think `category_account` and `category_content` are ever so slightly less confusing. But I don't have any better suggestion.

oneiros replied:
OK. I renamed the scopes.

ClearlyClaire commented on app/models/concerns/status/fasp_concern.rb:
```diff
@@ -0,0 +1,57 @@
+    return unless Mastodon::Feature.fasp_enabled?
+    return unless account_indexable? && public_visibility?
+
+    # We need the uri here, but it is set in another `after_commit`
+    # callback. Hooks included from modules are run before the ones
+    # in the class itself and can neither be reordered nor is there
+    # a way to declare dependencies.
+    store_uri if uri.nil?
```
An alternative, I think, is to do `prepend: true` in the `after_commit` in the main class. Not sure this really is better though.

ClearlyClaire replied:
That looks like to be the case indeed, good catch. This is a bit annoying.

ClearlyClaire commented on app/controllers/api/fasp/data_sharing/v0/backfill_requests_controller.rb:
```diff
@@ -0,0 +1,26 @@
+        if backfill_request.save
+          render json: { backfillRequest: { id: backfill_request.id } }, status: 201
+        else
+          head 422
```
Would it be useful to return the actual validation errors in the body?

ClearlyClaire commented on app/models/concerns/account/fasp_concern.rb:
Should these be called for all known accounts, or local ones only?

ClearlyClaire commented on app/models/concerns/status/fasp_concern.rb:
Should these be called for all known statuses, or only for local ones?

ClearlyClaire reviewed (commented):
This looks good overall, but I have added a few inline comments.

ClearlyClaire commented on app/models/concerns/status/fasp_concern.rb:
```diff
@@ -0,0 +1,57 @@
+    Fasp::AnnounceTrendWorker.perform_async(candidate_id, trend_source) if candidate_id
+  end
+
+  def public_visibility_or_just_changed?
+    public_visibility? || visibility_previously_was == 'public'
+  end
```
I guess it could be kept, although we currently have no mechanism at all to support visibility changes. A status being changed in the Rails console would not be properly reflected in the federation.

oneiros replied:
I decided to remove it to prevent any confusion.

oneiros committed:
Remove method as pointed out in review

Changing a status' visibility in an update should never happen.
Even if it does, this would not be federated, so this would
never be actually useful.

oneiros committed:
Rename scopes
